// @envelop/graphql-jit/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

const graphql = require('graphql');
const graphqlJit = require('graphql-jit');
const lru = _interopDefault(require('tiny-lru'));

const DEFAULT_MAX = 1000;
const DEFAULT_TTL = 3600000;
const useGraphQlJit = (compilerOptions = {}, pluginOptions = {}) => {
    const documentSourceMap = new WeakMap();
    const jitCache = typeof pluginOptions.cache !== 'undefined' ? pluginOptions.cache : lru(DEFAULT_MAX, DEFAULT_TTL);
    function getCacheEntry(args) {
        var _a;
        let cacheEntry;
        const documentSource = documentSourceMap.get(args.document);
        if (documentSource) {
            cacheEntry = jitCache.get(documentSource);
        }
        if (!cacheEntry) {
            const compilationResult = graphqlJit.compileQuery(args.schema, args.document, (_a = args.operationName) !== null && _a !== void 0 ? _a : undefined, compilerOptions);
            if (!graphqlJit.isCompiledQuery(compilationResult)) {
                if (pluginOptions === null || pluginOptions === void 0 ? void 0 : pluginOptions.onError) {
                    pluginOptions.onError(compilationResult);
                }
                else {
                    console.error(compilationResult);
                }
                cacheEntry = {
                    query: () => compilationResult,
                };
            }
            else {
                cacheEntry = compilationResult;
            }
            if (documentSource) {
                jitCache.set(documentSource, cacheEntry);
            }
        }
        return cacheEntry;
    }
    return {
        onParse({ params: { source } }) {
            const key = source instanceof graphql.Source ? source.body : source;
            return ({ result }) => {
                if (!result || result instanceof Error)
                    return;
                documentSourceMap.set(result, key);
            };
        },
        async onExecute({ args, setExecuteFn }) {
            if (!pluginOptions.enableIf || (pluginOptions.enableIf && (await pluginOptions.enableIf(args)))) {
                setExecuteFn(function jitExecutor() {
                    const cacheEntry = getCacheEntry(args);
                    return cacheEntry.query(args.rootValue, args.contextValue, args.variableValues);
                });
            }
        },
        async onSubscribe({ args, setSubscribeFn }) {
            if (!pluginOptions.enableIf || (pluginOptions.enableIf && (await pluginOptions.enableIf(args)))) {
                setSubscribeFn(async function jitSubscriber() {
                    const cacheEntry = getCacheEntry(args);
                    return cacheEntry.subscribe
                        ? cacheEntry.subscribe(args.rootValue, args.contextValue, args.variableValues)
                        : cacheEntry.query(args.rootValue, args.contextValue, args.variableValues);
                });
            }
        },
    };
};

exports.useGraphQlJit = useGraphQlJit;
